src/Sarif.Driver/Sdk/ExportConfigurationCommandBase.cs

```diff
@@ -20,17 +23,41 @@ public override int Run(ExportConfigurationOptions exportOptions)
 
                 // The export command could be updated in the future to accept an arbitrary set
                 // of analyzers for which to build an options XML file suitable for configuring them.
-                // Currently, we perform discovery against the built-in CodeFormatter rules
-                // and analyzers only.
                 ImmutableArray<IOptionsProvider> providers = DriverUtilities.GetExports<IOptionsProvider>(DefaultPlugInAssemblies);
                 foreach (IOptionsProvider provider in providers)
                 {
+                    IOption sampleOption = null;
+
+                    // Every analysis options provider has access to the following default configuration knobs
                     foreach (IOption option in provider.GetOptions())
                     {
+                        sampleOption = sampleOption ?? option;
                         allOptions.SetProperty(option, option.DefaultValue, cacheDescription: true);
                     }
                 }
 
+                IEnumerable<IRule> rules;
+                rules = DriverUtilities.GetExports<IRule>(DefaultPlugInAssemblies);
+
+                // This code injects properties that are provided for every rule instance.
+                foreach (IRule rule in rules)
+                {
+                    object objectResult;
+                    PropertiesDictionary properties;
+
+                    string ruleOptionsKey = rule.Id + "." + rule.Name + ".Options";
+
+                    if (!allOptions.TryGetValue(ruleOptionsKey, out objectResult))
+                    {
+                        objectResult = allOptions[ruleOptionsKey] = new PropertiesDictionary();
+                    }
+                    properties = (PropertiesDictionary)objectResult;
+
+                    foreach (IOption option in DefaultDriverOptions.Instance.GetOptions())
```

Comment:
Design:  So, this particular implementation seems to mean that consumers of the driver SDK cannot provide more default cross rule options, correct?  Do we want to enable that?  

I know the Driver+SDK+BinSkim are pretty tightly coupled, but this sort of thing will increase that coupling (esp. if down the line we want to add more "cross rule" configuration options for BinSkim's sake).

Just something to think about; I think this is fine for now, but we may want to add a comment in the DefaultDriverOptions or etc. about reducing this coupling in the future rather than adding more options there.

Reply:
Yes, right now other consumers cant enable this sort of thing, would need to be elevated to the SDK incrementally.,

Agreed, the coupling is clear and unfortunate. It's not a great way to evolve the driver framework, lots of assumptions that are relevant to binskim only are baked in...

---
In reply to: [169830521](https://github.com/Microsoft/sarif-sdk/pull/791#discussion_r169830521) [](ancestors = 169830521)